### DISCUSSION TITLE
Using ImmGetCompositionString yields a different result than Microsoft IME.

### DISCUSSION THREAD

**Discussion by changhao6** [google/mozc#858] [Q&A] [answered]
## Description
Hi, I am using the Windows API ImmGetCompositionString to get the result of IMECompositionEnd. When I call ImmGetCompositionString, Microsoft IME returns "アイウエオ" but Google IME returns "アイ". What is the reason for this result? How can I get the same result as using Microsoft IME? This is very important to me, please take a look at this problem in your busy schedule, thank you!

## Expected behavior
Here is MS IME behavior
![image](https://github.com/google/mozc/assets/47862827/0ed12019-48ee-4f5a-885a-bd02beb8a3bd)
get 'アイウエオ'

## Actual behavior
Here is Google IME behavior
![image](https://github.com/google/mozc/assets/47862827/a39068f5-9512-4b8f-ac54-c36e918c7855)
get 'アイ'

## Version or commit-id

Google IME 2.28.5050.0

## Environment

Windows 10-22H2

## Additional context
Here is the demo I used to reproduce this issue, it is based on winforms:
https://github.com/changhao6/IMEDemo

**Comment by yukawa** [comment]
What's exactly the steps you tried?  I actually got `アイウエオ` with the following steps.
1. Build and install the latest version of Mozc (c048e1fff6e84d1841c972a470d27ebac3f3f1c7 - you can download the installer file from [our GitHub Actions page](https://github.com/google/mozc/actions/runs/7318771942))
2. Build your demo app
3. Select Mozc as the current IME
4. Type `aiueo` to get `あいうえお`
5. Hit the Enter key to commit the composing text. <img width="227" alt="aiueo2" src="https://github.com/google/mozc/assets/5861397/09b25cc7-b297-46e0-8b93-a30060909242">

> Here is the demo I used to reproduce this issue, it is based on winforms:
> https://github.com/changhao6/IMEDemo

Most likely it's unrelated to this particular issue, but the following code would be a bit more accurate on how `ImmGetCompositionString` should be used from C#. Just fyi.

```
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace IME
{
    internal class MyTextBox : TextBox
    {

        const int GCS_RESULTREADSTR = 0x0200;
        public const int WM_IME_COMPOSITION = 0x010F;
        public const int WM_IME_ENDCOMPOSITION = 0x010E;

        protected override void WndProc(ref Message m)
        {
            if (m.Msg == WM_IME_ENDCOMPOSITION)
            {
                MessageBox.Show(GetCompositionString(this, GCS_RESULTREADSTR));
            }
            base.WndProc(ref m);
        }

        static string GetCompositionString(TextBox textBox, uint dwIndex)
        {
            IntPtr hWnd = textBox.Handle;
            IntPtr hIMC = IntPtr.Zero;
            try
            {
                hIMC = ImmGetContext(hWnd);
                int dwSize = ImmGetCompositionString(hIMC, dwIndex, null, 0);
                if (dwSize > 0)
                {
                    byte[] buf = new byte[dwSize];
                    int copiedSize = ImmGetCompositionString(hIMC, dwIndex, buf, dwSize);
                    if (copiedSize >= 0)
                    {
                        return Encoding.Unicode.GetString(buf, 0, dwSize);
                    }
                }
                return string.Empty;
            }
            finally
            {
                if (hIMC != IntPtr.Zero)
                {
                    ImmReleaseContext(hWnd, hIMC);
                }
            }
        }

        [DllImport("imm32.dll", CharSet = CharSet.Unicode)]
        static extern int ImmGetCompositionString(
            IntPtr hIMC,
            uint dwIndex,
            [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 3)]
            byte[] lpBuf,
            int dwBufLen);
        [DllImport("imm32.dll")]
        public static extern IntPtr ImmGetContext(IntPtr hWnd);
        [DllImport("imm32.dll")]
        public static extern IntPtr ImmReleaseContext(IntPtr hWnd, IntPtr hIMC);
    }
}
```

**Reply by changhao6** [reply]
I'm sorry,I didn't make that clear. In fact, I just typed `ai`, and  chose 'あいうえお' in candidate list. When I performed the same operation on Microsoft IME, I can get the whole 'ｱｲｳｴｵ'. 
![ai](https://github.com/google/mozc/assets/47862827/9c12068e-8146-47ea-b9fd-60f82e164052)

**Reply by changhao6** [reply]
@yukawa ,I had ran your modified code, but the problem still exists. Could you please take some time to look at it again?

**Reply by yukawa** [reply, marked as answer]
OK, I can finally reproduce the issue you described.

> How can I get the same result as using Microsoft IME? This is very important to me, please take a look at this problem in your busy schedule, thank you!

The short answer to your original question is that unfortunately there is no way to get the same result as MS-IME unless we change the behavior of Mozc / Google Japanese Input. The behavior would be much similar as long as you use the space key to convert text instead of the down arrow key. As long as the user always uses the space key to convert the text, the issue is less likely to happen.

The long answer is that when APIs like `ImmGetCompositionString` were defined, prefix match suggestion / prediction wasn't popular in Japanese IMEs, and in fact MS-IME always yielded exact match results as if you hit the space key. We released Mozc/Google Japanese Input in the current manner (setting `アイ` as a reading string for prefix suggestion) in 2009, followed by MS-IME's starting supporting prefix match suggestion / prediction in their way (setting `アイウエオ` as a reading string based on the result string).

For this particular case, I think it'd make sense to update Mozc's behavior to match the MS-IME's behavior, but we cannot promise whether/when it would happen. I'll update this thread when we update Mozc's behavior to match the MS-IME's behavior.

Hope this helps.

**Reply by changhao6** [reply]
Thanks for your reply. Your answer has helped me a lot!
